### ISSUE TITLE
Initialize random seed at the beginning of an FBPIC simulation

### ISSUE THREAD

**Issue by RemiLehe** [fbpic/fbpic#48]
Initializing the random seed (on the CPU, as well as GPU when used) would make the simulations exactly reproducible.

As a reminder, the random number generator is currently used for:
- determining the angular position of the particles of the plasma, when they are initialized
- determining whether an ionization event will occur.

**Comment by RemiLehe** [comment]
Note: we should probably at least indicate why the results of identical simulations can fluctuate, in the documentation.

**Comment by soerenjalas** [comment]
Would it make sense to use a fixed seed for the random number generator to have consistent results of two equivalent simulations?

**Comment by RemiLehe** [comment]
Yes, this would definitely make sense, but there are a few caveat:
- Setting the random seed with `np.random.seed` would fix the random angles at which particles are initialized. But it would not affect the random generator used for ionization calculations (which does not use numpy directly, but is instead automatically created by `numba`/`numba.cuda`).
- During a restart, it is not obvious how to retrieve the state in which a random generator was, at the time of the checkpoint. So a restarted simulation would probably still give different results than a full simulation. (Unless numpy provides an API for getting the current random state, and we then save it in the checkpoint.)

**Comment by RemiLehe** [comment]
Also, if we use `np.random.seed`, I am not sure when to call this:
- When FBPIC is imported ?
- When the `Simulation` object is created? (but then initializing several `Simulation` in the same script object would overwrite the seed.)

**Comment by RemiLehe** [comment]
Also, if we decide to set the seed when the `Simulation` object is created, we could add the seed as an argument of the `Simulation` object, which defaults to `None` (i.e. when this is `None`, the seed is not set and the results will fluctuate from one run to the next.)

**Comment by abonatto** [comment]
Hi @RemiLehe,

Thank you very much for your fast reply.

Sure thing. Let me install and test the version you pointed me out, and I will let you know the outcome. Hopefully the small fluctuations you mentioned will not be relevant for our case.

**Comment by abonatto** [comment]
Hi @RemiLehe,

I just installed it (pip install git+https://github.com/fbpic/fbpic.git@dev). In our current script, in order to set the CPU seed, we set the numpy seed (for example, np.random.seed(0)) within the simulation routine, after the "sim = ..." and before adding the particle species. Regarding the GPU seed, could you please point me out how can I set it? In addition, if I am using four GPUs, I suppose I should set a (distinct) seed for each of them, correct?

**Comment by RemiLehe** [comment]
Hi @abonatto 
In order to install the code from the new PR (https://github.com/fbpic/fbpic/pull/599), you will need to do:
```
pip install git+https://github.com/RemiLehe/fbpic@fix_random_seed
```
Once you do this, you will not need to set `np.random.seed` in your input script anymore, but instead you would pass `random_seed=0` directly to the `Simulation` object (since the code changes in the PR). This will also automatically set the seed for the GPU code. 
Regarding setting a different seed for each MPI ranks: this would be a good idea, but I don't think it is crucial at this point. For now, if you just want to confirm that simulations are reproducible, you can simply set `random_seed=0` and evaluate whether you get the same results when running the same simulation twice.

**Comment by abonatto** [comment]
Hi @RemiLehe,

Thanks for your detailed explanation. I will run the simulations using the new PR, and let you know the results.

**Comment by OwenSlevert** [comment]
Hello, I would have a short question about those seeds. I am currently simulating laser wakefield acceleration without any ionization. With a fixed seed, I am still observing rather different injected charges. Is this still in the scope when setting the seed at the beginning or do I have to do anything more? Are there any random numbers not affected by setting the seed (besides ionization)? I am currently working only with one thread on one GPU. Thank you very much!

**Comment by MKirchen** [comment]
Did you set the seed with the corresponding helper function as described [here](https://fbpic.github.io/api_reference/simulation.html)?

**Comment by OwenSlevert** [comment]
Yes I did. I saw that for the first 12000 timesteps we can reproduce the simulations, but then we see electrons beeing at different positions compared to each other. But, when using less modes it seems not so significant compared to using 6 modes, but this I am still looking into.
